src/main/cpp/src/cast_string_to_timestamp.cu

```diff
@@ -0,0 +1,876 @@
+
+  // Not specified timezone in the string, indicate to use the default timezone.
+  NOT_SPECIFIED = 0,
+
+  // Fixed offset timezone
+  // String starts with UT/GMT/UTC/[+-], and it's valid.
+  // E.g: +08:00, +08, +1:02:30, -010203, GMT+8, UTC+8:00, UT+8
+  // E.g: +01:2:03
+  FIXED_TZ = 1,
+
+  // Not FIXED_TZ, it's a valid timezone string.
+  // E.g.: java.time.ZoneId.SHORT_IDS: CTT
+  // E.g.: Region-based timezone: America/Los_Angeles
+  OTHER_TZ = 2,
+
+  // Invalid timezone.
+  // String starts with UT/GMT/UTC/[+-], but it's invalid.
+  // E.g: UTC+19:00, GMT+19:00, max offset is 18 hours
+  // E.g: GMT+01:2:03, +01:2:03, special case
+  // E.g: non-exist-timezone
+  INVALID_TZ = 3
+};
+
+/**
+ * Represents a timezone.
+ */
+struct time_zone {
+  TZ_TYPE type;
+
+  // for fixed timezone
+  int fixed_offset;
+
+  // for java.time.ZoneId.SHORT_IDS and Region-based zone IDs
+  // the tz offset to the input string
+  int tz_pos_in_string;
+
+  // for java.time.ZoneId.SHORT_IDS and Region-based zone IDs
+  // the tz end offset to the input string
+  int tz_end_pos_in_string;
+
+  __device__ int tz_len() const { return tz_end_pos_in_string - tz_pos_in_string; }
+
+  __device__ time_zone()
+    : type(TZ_TYPE::NOT_SPECIFIED), fixed_offset(0), tz_pos_in_string(0), tz_end_pos_in_string(0)
+  {
+  }
+  __device__ time_zone(TZ_TYPE t, int offset, int tz_pos, int tz_end_pos)
+    : type(t), fixed_offset(offset), tz_pos_in_string(tz_pos), tz_end_pos_in_string(tz_end_pos)
+  {
+  }
+};
+
+__device__ time_zone make_fixed_tz(int offset)
+{
+  return time_zone(TZ_TYPE::FIXED_TZ, offset, 0, 0);
+}
+
+__device__ time_zone make_invalid_tz() { return time_zone(TZ_TYPE::INVALID_TZ, 0, 0, 0); }
+
+__device__ time_zone make_other_tz(int tz_pos, int tz_end_pos)
+{
+  return time_zone(TZ_TYPE::OTHER_TZ, 0, tz_pos, tz_end_pos);
+}
+
+/**
+ * Is white space, consistent with Spark UTF8String.trimAll for char input
+ */
+__device__ bool is_whitespace(char const c)
+{
+  // Keep consistent with Java: Character.isWhitespace(c) ||
+  // Character.isISOControl(c),
+  // 0-31 is control characters, 32 is space, 127 is delete
+  return (c >= 0 && c <= 32) || c == 127;
+}
+
+__device__ bool is_valid_tz(time_zone const& tz) { return tz.type != TZ_TYPE::INVALID_TZ; }
+
+/**
+ * Parse a string to an integer.
+ * @param ptr The input string
+ * @param[out] pos the pointer to the string when parsing
+ * @param end_pos the end pointer of the string
+ * @param[out] v the parsed value
+ * @param min_digits the minimum digits to parse, if not enough, return false
+ * @param max_digits the maximum digits to parse, if exceeds, return false
+ * @return true if the input string is valid, false otherwise
+ */
+__device__ bool parse_int(
+  char const* const ptr, int& pos, int const end_pos, int& v, int min_digits, int max_digits)
+{
+  v          = 0;
+  int digits = 0;
+  while (pos < end_pos) {
+    int const parsed_value = static_cast<int32_t>(ptr[pos] - '0');
```

Comment:
`char` type is not guaranteed to be signed. It can be unsigned thus we must cast to `int` before doing subtraction.
```suggestion
    int const parsed_value = static_cast<int32_t>(ptr[pos]) - '0';
```